### PR TITLE
[FLINK-19229][python] Introduce the PythonStreamGroupAggregateOperator for Python UDAF.

### PR DESCRIPTION
**Pull request by WeiZhong94**
## What is the purpose of the change

*This pull request introduces the PythonStreamGroupAggregateOperator for Python UDAF.*

## Brief change log

  - *Add stateful runners for executing Python UDAF via Beam Fn Execution Framework.*
  - *Introduce the `PythonStreamGroupAggregateOperator` for Python UDA*

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/utils/OperatorUtils.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import org.apache.flink.table.runtime.types.CRow;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+
+import com.google.protobuf.ByteString;
+
+/**
+ * The collectors used to collect Row values.
+ */
+public enum OperatorUtils {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/utils/OperatorUtils.java:34]: Rename to PythonOperatorUtils?

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your comments! I have updated this PR.

File: flink-python/pyflink/proto/flink-fn-execution.proto [hunk 2]
```diff
@@ -71,6 +71,24 @@ message UserDefinedDataStreamFunctions {
   bool metric_enabled = 2;
 }
 
+// A list of the user-defined aggregate functions to be executed in an group aggregate operation.
```
Comment (dianfu) [inline flink-python/pyflink/proto/flink-fn-execution.proto:74]:
```suggestion
// A list of the user-defined aggregate functions to be executed in a group aggregate operation.
```

File: flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamPythonFunctionRunner.java [hunk 3]
```diff
@@ -41,30 +60,60 @@
...
 import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
 import java.util.concurrent.LinkedBlockingQueue;
 
+import static org.apache.beam.runners.core.construction.BeamUrns.getUrn;
+
 /**
- * An base class for {@link PythonFunctionRunner} based on beam.
+ * A {@link BeamPythonFunctionRunner} used to execute Python stateful functions.
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamPythonFunctionRunner.java:95]:
```suggestion
 * A {@link BeamPythonFunctionRunner} used to execute Python functions.
```

File: flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamPythonFunctionRunner.java [hunk 4]
```diff
@@ -41,30 +60,60 @@
...
  */
 @Internal
 public abstract class BeamPythonFunctionRunner implements PythonFunctionRunner {
 	protected static final Logger LOG = LoggerFactory.getLogger(BeamPythonFunctionRunner.class);
 
-	private transient boolean bundleStarted;
+	private static final String BEAM_STATE_PREFIX = "beam-state-";
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamPythonFunctionRunner.java:101]: What about `python-state-`?

File: flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamPythonFunctionRunner.java [hunk 5]
```diff
@@ -309,4 +356,287 @@ private void checkInvokeStartBundle() throws Exception {
...
+	private static class SimpleStateRequestHandler implements StateRequestHandler {
+
+		private final TypeSerializer keySerializer;
+		private final TypeSerializer<byte[]> valueSerializer;
+		private final KeyedStateBackend keyedStateBackend;
+
+		/**
+		 * Reusable OutputStream used to holding the serialized input elements.
+		 */
+		private final ByteArrayOutputStreamWithPos baos;
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamPythonFunctionRunner.java:490]: Can be converted to a local variable.

File: flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamPythonFunctionRunner.java [hunk 6]
```diff
@@ -309,4 +356,287 @@ private void checkInvokeStartBundle() throws Exception {
...
+
+		/**
+		 * Reusable InputStream used to holding the elements to be deserialized.
+		 */
+		private final ByteArrayInputStreamWithPos bais;
+
+		/**
+		 * OutputStream Wrapper.
+		 */
+		private final DataOutputViewStreamWrapper baosWrapper;
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamPythonFunctionRunner.java:500]: It's never used?

File: flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamPythonFunctionRunner.java [hunk 7]
```diff
@@ -309,4 +356,287 @@ private void checkInvokeStartBundle() throws Exception {
...
+		 */
+		final Map<String, StateDescriptor> stateDescriptorCache;
+
+		SimpleStateRequestHandler(
+			KeyedStateBackend keyedStateBackend,
+			TypeSerializer keySerializer) {
+			this.keySerializer = keySerializer;
+			this.valueSerializer = PrimitiveArrayTypeInfo.BYTE_PRIMITIVE_ARRAY_TYPE_INFO
+				.createSerializer(new ExecutionConfig());
+			this.keyedStateBackend = keyedStateBackend;
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamPythonFunctionRunner.java:518]: Move this line to the front of the constructor?

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/RowDataPythonStreamGroupAggregateOperator.java [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.table.runtime.typeutils.PythonTypeUtils.toProtoType;
+
+/**
+ *  The Python AggregateFunction operator for the blink planner.
+ */
+public class RowDataPythonStreamGroupAggregateOperator
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/RowDataPythonStreamGroupAggregateOperator.java:72]: Remove the RowData prefix as we will not support the old planner any more and so the prefix is not necessary.

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/RowDataPythonStreamGroupAggregateOperator.java [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+
+	@VisibleForTesting
+	protected static final String STREAM_GROUP_AGGREGATE_URN = "flink:transform:stream_group_aggregate:v1";
+
+	@VisibleForTesting
+	protected static final byte NORMAL_RECORD = 0;
+
+	private static final byte TRIGGER_TIMER = 1;
+
+	private static final long serialVersionUID = 1L;
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/RowDataPythonStreamGroupAggregateOperator.java:92]: Move this line to the front of the class?

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/RowDataPythonStreamGroupAggregateOperator.java [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+
+	/**
+	 * Generate retract messages if true.
+	 */
+	private final boolean generateUpdateBefore;
+
+	/**
+	 * The minimum time in milliseconds until state which was not updated will be retained.
+	 */
+	final long minRetentionTime;
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/RowDataPythonStreamGroupAggregateOperator.java:131]: private?

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/RowDataPythonStreamGroupAggregateOperator.java [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+
+	/**
+	 * The maximum time in milliseconds until state which was not updated will be retained.
+	 */
+	final long maxRetentionTime;
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/RowDataPythonStreamGroupAggregateOperator.java:136]: ditto

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/RowDataPythonStreamGroupAggregateOperator.java [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+
+	/**
+	 * The maximum NUMBER of the states cached in Python side.
+	 */
+	private int stateCacheSize = DEFAULT_STATE_CACHE_SIZE;
+
+	/**
+	 * Indicates whether state cleaning is enabled. Can be calculated from the `minRetentionTime`.
+	 */
+	protected final boolean stateCleaningEnabled;
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/RowDataPythonStreamGroupAggregateOperator.java:146]: can be private

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/RowDataPythonStreamGroupAggregateOperator.java [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+		userDefinedFunctionInputType = new RowType(fields);
+		udfInputTypeSerializer = PythonTypeUtils.toBlinkTypeSerializer(userDefinedFunctionInputType);
+		udfOutputTypeSerializer = PythonTypeUtils.toBlinkTypeSerializer(outputType);
+		bais = new ByteArrayInputStreamWithPos();
+		baisWrapper = new DataInputViewStreamWrapper(bais);
+		baos = new ByteArrayOutputStreamWithPos();
+		baosWrapper = new DataOutputViewStreamWrapper(baos);
+		timerService = new SimpleTimerService(
+			getInternalTimerService("state-clean-timer", VoidNamespaceSerializer.INSTANCE, this));
+		reuseRowData = new UpdatableRowData(GenericRowData.of(NORMAL_RECORD, null, null, null), 4);
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/RowDataPythonStreamGroupAggregateOperator.java:252]: Add some comments about the meaning of each column?

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/RowDataPythonStreamGroupAggregateOperator.java [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+			cleanupTimeState = getRuntimeContext().getState(inputCntDescriptor);
+		}
+	}
+
+	private Map<String, String> buildJobOptions(Configuration config) {
+		Map<String, String> jobOptions = new HashMap<>();
+		if (config.containsKey("table.exec.timezone")) {
+			jobOptions.put("table.exec.timezone", config.getString("table.exec.timezone", null));
+		}
+		if (config.containsKey("python.state.cache.size")) {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/RowDataPythonStreamGroupAggregateOperator.java:352]: How users could know about this config? Should we add it in PythonConfig?

**Review by dianfu — COMMENTED** [review]
@WeiZhong94  Thanks a lot for the update.

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your comments! I have updated this PR.

**Review by dianfu — APPROVED** [review]
LGTM

**Pull request merged**